### PR TITLE
Customizable serialization for `SWCurveConfig` and `TECurveConfig`

### PR DESCRIPTION
**Pull request by alexander-zw**
## Description

closes: #308

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: ec/src/models/short_weierstrass/affine.rs [hunk 1]
```diff
@@ -125,7 +124,7 @@ impl<P: SWCurveConfig> Affine<P> {
     ///
     /// The results are sorted by lexicographical order.
     /// This means that, if `P::BaseField: PrimeField`, the results are sorted as integers.
-    fn get_ys_from_x_unchecked(x: P::BaseField) -> Option<(P::BaseField, P::BaseField)> {
+    pub fn get_ys_from_x_unchecked(x: P::BaseField) -> Option<(P::BaseField, P::BaseField)> {
```
Comment (alexander-zw) [inline ec/src/models/short_weierstrass/affine.rs:127]: Deserialize uses `get_ys_from_x_unchecked`. Should I move `get_ys_from_x_unchecked` and `get_point_from_x_unchecked` to the curve config too? If not, I need to make this public.

File: ec/src/models/short_weierstrass/affine.rs [hunk 2]
```diff
@@ -344,38 +343,18 @@ impl<P: SWCurveConfig> From<Projective<P>> for Affine<P> {
 }
 
 impl<P: SWCurveConfig> CanonicalSerialize for Affine<P> {
-    #[allow(unused_qualifications)]
     #[inline]
     fn serialize_with_mode<W: Write>(
         &self,
-        mut writer: W,
```
Comment (alexander-zw) [inline ec/src/models/short_weierstrass/affine.rs:351]: TIL that you can turn an immutable variable into a mutable one

File: ec/src/models/short_weierstrass/affine.rs [hunk 3]
```diff
@@ -344,38 +343,18 @@ impl<P: SWCurveConfig> From<Projective<P>> for Affine<P> {
 }
 
 impl<P: SWCurveConfig> CanonicalSerialize for Affine<P> {
-    #[allow(unused_qualifications)]
```
Comment (alexander-zw) [inline ec/src/models/short_weierstrass/affine.rs:347]: I'm not sure what this does, but there are no lint errors after removing it. Let me know if it's wrong to remove

File: ec/src/models/short_weierstrass/affine.rs [hunk 4]
```diff
@@ -390,47 +369,12 @@ impl<P: SWCurveConfig> Valid for Affine<P> {
 }
```
Comment (alexander-zw) [inline ec/src/models/short_weierstrass/affine.rs:369]: Should I move `check()` into the curve config too?

File: ec/src/models/twisted_edwards/mod.rs [hunk 5]
```diff
@@ -78,6 +84,63 @@ pub trait TECurveConfig: super::CurveConfig {
...
+        }
+        Ok(point)
+    }
+
+    #[inline]
+    fn serialized_size(compress: Compress) -> usize {
+        let zero = Self::BaseField::zero();
+        match compress {
+            Compress::Yes => zero.serialized_size_with_flags::<TEFlags>(),
+            Compress::No => zero.uncompressed_size() + zero.uncompressed_size(),
```
Comment (alexander-zw) [inline ec/src/models/twisted_edwards/mod.rs:141]: I changed this from `x.uncompressed_size() + y.uncompressed_size()` to `zero.uncompressed_size() + zero.uncompressed_size()`. Is there a difference? Aren't they all the same field? With this change this function is independent of the affine element.

File: ec/src/models/short_weierstrass/affine.rs (hunk 3 above)
Comment (alexander-zw) [reply ec/src/models/short_weierstrass/affine.rs:347]: I dug into the history and it's been here for at least three years, I'm guessing it was kept long after no longer being needed

File: ec/src/models/twisted_edwards/mod.rs (hunk 5 above)
Comment (alexander-zw) [reply ec/src/models/twisted_edwards/mod.rs:141]: The tests passed after this change, though I'm not sure how well serialize size is tested

File: ec/src/models/short_weierstrass/affine.rs (hunk 1 above)
Comment (Pratyush) [reply ec/src/models/short_weierstrass/affine.rs:127]: You can make it `pub`

File: ec/src/models/short_weierstrass/affine.rs (hunk 3 above)
Comment (Pratyush) [reply ec/src/models/short_weierstrass/affine.rs:347]: Yeah, we can remove it if it doesn't trigger any linter errors.

File: ec/src/models/short_weierstrass/affine.rs (hunk 4 above)
Comment (Pratyush) [reply ec/src/models/short_weierstrass/affine.rs:369]: I think for the time being we can keep it outside; for elliptic curves, the checks are basically `is_in_subgroup_assuming_is_on_curve`, and we already specialize that.

File: ec/src/models/twisted_edwards/mod.rs (hunk 5 above)
Comment (Pratyush) [reply ec/src/models/twisted_edwards/mod.rs:141]: It's the same right now, but if it's simpler we can keep the old behaviour.

Comment (alexander-zw) [reply ec/src/models/twisted_edwards/mod.rs:141]: The new behavior is simpler. In the new behavior the size doesn't depend on the curve element, in the old behavior it referenced the curve values. So should we keep the behavior in this PR right now?

**Review by Pratyush — APPROVED** [review]

**Pull request merged**